### PR TITLE
Resource Elasticity [2/3]

### PR DESCRIPTION
**Pull request by mbrodmann**
This PR is the second of three planned PRs for implementing resource elasticity within Crail. More specifically, the goal is to allow to automatically and dynamically adapt the number of running datanodes based on different possible parameters (e.g. resource consumption, available storage capacities, ...). This requires support for starting and terminating running datanodes automatically. This functionality will be implemented mostly within the Crail namenode. Therefore, the Crail namenode will not only maintain information on the current state of the datanodes and the deployment but will also start or terminate datanodes as required. For deciding how to adjust the number of datanodes  based on the current state observed in the Crail namenode so-called _policies_ are used. 

The following changes will be included within the seperate PRs:

1) The first PR implements a new RPC (for both darpc and narpc) that allows to specify a datanode (using IP-address and port number) that should be shutdown.

2) In this second PR the actual mechanism for shutting down a datanode is covered. This includes checking that a datanode can be removed safely (i.e. the datanode does not store any remaining data), otherwise it will only be scheduled for removal. This PR also handles the datanode termination for all of the three available storage layers. 

3) The third PR implements a simple example policy for implementing elasticity based on the system's storage consumption.

### CODE REVIEW

File: namenode/src/main/java/org/apache/crail/namenode/BlockStore.java [hunk 1]
```diff
@@ -171,9 +192,33 @@ short addDataNode(DataNodeBlocks dataNode) {
...
+
+	short removeDatanode(DataNodeInfo dn) throws Exception {
+		// this will remove it as well
+		return _prepareOrRemoveDN(dn, false);
+	}
+
 	//---------------
-	
+
+	private short _prepareOrRemoveDN(DataNodeInfo dn, boolean onlyMark) throws Exception {
```
Comment (patrickstuedi) [inline namenode/src/main/java/org/apache/crail/namenode/BlockStore.java:208]: I'm not a big fan of using "_" for private methods, maybe just rename the methods of this type

**Review by patrickstuedi — CHANGES_REQUESTED** [review]

**Comment by PepperJo** [comment]
Can you rebase this PR on the latest master

**Comment by mbrodmann** [comment]
> Can you rebase this PR on the latest master

The branch should now be rebased on the latest master

File: namenode/src/main/java/org/apache/crail/namenode/BlockStore.java [hunk 2]
```diff
@@ -171,9 +192,33 @@ short addDataNode(DataNodeBlocks dataNode) {
...
-	
+
+	private short prepareOrRemoveDN(DataNodeInfo dn, boolean onlyMark) throws Exception {
+		DataNodeBlocks toBeRemoved = membership.get(dn.key());
+		if (toBeRemoved == null) {
+			LOG.error("DataNode: " + dn.toString() + " not found");
+			return RpcErrors.ERR_DATANODE_NOT_REGISTERED;
+		} else {
+			if (onlyMark)
+				toBeRemoved.scheduleForRemoval();
```
Comment (PepperJo) [inline namenode/src/main/java/org/apache/crail/namenode/BlockStore.java:215]: please use `{` also for single line if/else. It should be like this elsewhere.

File: namenode/src/main/java/org/apache/crail/namenode/BlockStore.java [hunk 3]
```diff
@@ -171,9 +192,33 @@ short addDataNode(DataNodeBlocks dataNode) {
...
+			else
+				membership.remove(toBeRemoved.key());
```
Comment (PepperJo) [inline namenode/src/main/java/org/apache/crail/namenode/BlockStore.java:217]: Can you explain what happens when we do not mark but just remove?

**Review by PepperJo — CHANGES_REQUESTED** [review]
Mostly cosmetics.

File: namenode/src/main/java/org/apache/crail/namenode/BlockStore.java (hunk 3 above)
Comment (mbrodmann) [reply namenode/src/main/java/org/apache/crail/namenode/BlockStore.java:218]: A datanode is only marked for removal first because it might still store datablocks. Only when no data is stored anymore by a marked datanode the namenode can start the actual removal process, that is removing it from its datastructures and notifying the datanode that it can terminate (this is periodically checked as part of the `getDataNode` RPC). When we just remove that datanode the data it stores is not available anymore.

Comment (PepperJo) [reply namenode/src/main/java/org/apache/crail/namenode/BlockStore.java:218]: So in the case the datanode still has data and is forcefully removed, we would get file with wholes in them, correct? Blocks of this datanodes that are still used in files are not updated that the datanode has been removed? So if a client asks for a block he will get the datanode info of a non-existing datanode? Worse if someone starts a datanode with the same ip/port at a later point he might see different data (depending on datanode type).

Comment (mbrodmann) [reply namenode/src/main/java/org/apache/crail/namenode/BlockStore.java:218]: I'm not sure I fully understand the problem you are referring to. Could you rephrase perhaps?
My understanding is that when a datanode is forcefully removed (also independently of this PR, e.g. by killing it) we run into the problem you mentioned that certain blocks of files might be missing. When a client requests a file the namenode will point to the corresponding datanodes. As the datanode was forcefully shutdown the client application will not be able to retrieve the missing blocks. Am I missing a point here?
To prevent this situation the idea was to only allow the (actual) removal of a datnode when it does not store any remaining blocks.

Comment (PepperJo) [reply namenode/src/main/java/org/apache/crail/namenode/BlockStore.java:218]: Sorry for the confusion I was not referring to killing a datanode. Is this method always called first with `onlyMark=true` (e.g. prepare) and later you remove or is there a case where one can "force" remove and call with `onlyMark=false` directly?
I think given that this method does two different things we should split it in a prepare and a remove method.

Comment (mbrodmann) [reply namenode/src/main/java/org/apache/crail/namenode/BlockStore.java:218]: Thank you for clarifying. Yes, that is correct a datanode is always marked first and removed afterwards when there are no remaining blocks (i.e. there is no "force" case).
I agree splitting this into two methods would make things clearer and avoid possible confusions. I will change it.

Comment (PepperJo) [reply namenode/src/main/java/org/apache/crail/namenode/BlockStore.java:218]: Thanks!

Comment (mbrodmann) [reply namenode/src/main/java/org/apache/crail/namenode/BlockStore.java]: I hope the new organization of the methods makes more sense.

File: client/src/main/java/org/apache/crail/tools/RemoveDataNode.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        HelpFormatter formatter = new HelpFormatter();
+        CommandLineParser parser = new DefaultParser();
+        CommandLine line = parser.parse(options, args);
+
+        try {
+            ipaddr = InetAddress.getByName(line.getOptionValue(ipOption.getOpt()));
+            port = Integer.parseInt(line.getOptionValue(portOption.getOpt()));
+        } catch(Exception e) {
+            formatter.printHelp("RemoveDataNode", options);
```
Comment (PepperJo) [inline client/src/main/java/org/apache/crail/tools/RemoveDataNode.java:38]: We should exit here, e.g. with System.exit(-1); or by rethrowing the exception

**Review by PepperJo — COMMENTED** [review]
One last thing. The rest looks good to me.

**Comment by mbrodmann** [comment]
> One last thing. The rest looks good to me.

Ok great! Thank you 👍

**Review by PepperJo — APPROVED** [review]

**Review by BernardMetzler — APPROVED** [review]
Looks good to me!
